code/modules/mob/living/carbon/xenomorph/Xenomorph.dm

```diff
@@ -117,40 +84,32 @@
 
 		nicknumber = tempnumber
 
-
-	//Larvas have their own, very weird naming conventions, let's not kick a beehive, not yet
-	if(isxenolarva(src))
-		return
-
-	var/name_prefix = ""
-
-	var/datum/hive_status/hive
-	if(hivenumber && hivenumber <= hive_datum.len)
-		hive = hive_datum[hivenumber]
-	else
-		hivenumber = XENO_HIVE_NORMAL
-		hive = hive_datum[hivenumber]
-
-	name_prefix = hive.prefix
-	color = hive.color
-	if(name_prefix == "Corrupted ")
-		add_language("English")
-	else
-		remove_language("English") // its hacky doing it here sort of
-
-	//Queens have weird, hardcoded naming conventions based on upgrade levels. They also never get nicknumbers
-	if(isxenoqueen(src))
-		switch(upgrade)
-			if(0) name = "[name_prefix]Queen"			 //Young
-			if(1) name = "[name_prefix]Elder Queen"	 //Mature
-			if(2) name = "[name_prefix]Elder Empress"	 //Elder
-			if(3) name = "[name_prefix]Ancient Empress" //Ancient
-	else name = "[name_prefix][xeno_caste.upgrade_name] [xeno_caste.display_name] ([nicknumber])"
+//Off-load this proc so it can be called freely
+//Since Xenos change names like they change shoes, we need somewhere to hammer in all those legos
+//We set their name first, then update their real_name AND their mind name
+/mob/living/carbon/Xenomorph/proc/generate_name()
+	name = "[hive.prefix][xeno_caste.upgrade_name] [xeno_caste.display_name] ([nicknumber])"
```

Comment:
I take it no xenos should ever be without a hive, right? Else this would runtime.

Reply:
Yeah they always need a hive and thats sanity checked before this